File: 3/sol.py
from sys import stdin
import re

text = "".join(stdin)

def part1(text):
    regex = r"mul\(\d{1,3},\d{1,3}\)"
    matches = [m for m in re.finditer(regex, text, re.MULTILINE)]
    s = 0
    for m in matches:
        a, b = map(int, m.group()[4:-1].split(","))
        s += a * b
    return s

def part2(text):
    regex = r"mul\(\d{1,3},\d{1,3}\)|do\(\)|don't\(\)"
    matches = [m for m in re.finditer(regex, text, re.MULTILINE)]
    on = True
    s = 0
    for m in matches:
        instr = m.group()
        if instr == "do()":
            on = True
        elif instr == "don't()":
            on = False
        else:
            a, b = map(int, instr[4:-1].split(","))
            s += on * a * b
    return s

print(f"part 1 is {part1(text)}")
print(f"part 2 is {part2(text)}")

File: 3/sol.rs
// cargo-deps: regex="1.11.1"
extern crate regex;

use std::io::{self, BufRead};
use regex::Regex;

fn part1(text: &str) -> usize {
    let re = Regex::new(r"mul\((\d{1,3}),(\d{1,3})\)").unwrap();
    return re
        .captures_iter(text)
        .map(|caps| {
            return caps
                .extract::<2>().1
                .iter()
                .map(|n| n.parse::<usize>().unwrap()) 
                .reduce(|a, b| a*b)
                .unwrap()
        })
        .sum::<usize>();
}

fn part2(text: &str) -> usize {
    let re = Regex::new(r"mul\((\d{1,3}),(\d{1,3})\)|do\(\)|don't\(\)").unwrap();    
    let mut res = 0;
    let mut on = 1;
    for cap in re.captures_iter(text) {
        let m = &cap[0];
        if m == "do()" {
            on = 1;
        } else if m == "don't()" {
            on = 0;
        } else {
            let fst = &cap[1].parse::<usize>().unwrap(); 
            let snd = &cap[2].parse::<usize>().unwrap(); 
            res += fst * snd * on;
        }
    }
    return res;
}

fn main() -> io::Result<()> {
    let stdin = io::stdin();
    let text = stdin
        .lock()
        .lines()
        .map(|line_res| line_res.unwrap())
        .collect::<String>();

    println!("part 1 is {}", part1(&text)); 
    println!("part 2 is {}", part2(&text)); 

    Ok(())
}
